### PR TITLE
[KTIR Emitter] Improve the emission surface and support sum on-stick

### PR DESCRIPTION
**Pull request by fabianlim**
> This rebases @tnakaike's #3859 on top 

Separates **what a compute op is** from **which `linalg` op spells it**, and uses that separation to emit a reduction along the stick — the shape [dataflow-scheduler#37](https://github.com/torch-spyre/dataflow-scheduler/issues/37) asks for.

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [x] cleanup

#### What this PR does:

`torch.sum(x, dim=-1)` folds the 64 lanes of a stick into one number, and the answer still occupies a whole stick on the way out. So the *same* lane axis is read as one dimension and written as another — one axis that is both reduced and kept.

`ComputeStep.reduce_dims` could not say that. It was a flat tuple of reduced input axes, computed as the set complement of the kept ones, so an axis was necessarily one or the other. The refusal a reduction of this shape actually hit was `"the surviving axes are permuted, which needs a transpose rather than a reduction"` — the wrong diagnosis, since nothing is transposed.

Three facts were tangled in `Family`, which classified the op *and* named the method that emitted it (reached by `getattr(self, step.family.name.lower())`). They are now separate:

| fact | where it lives | values |
|---|---|---|
| what `binding()` is, and how to call it | `Recipe.kind` | `NAMED` (`linalg.add`) / `PAYLOAD` (a scalar for a parallel body) / `COMBINER` (`arith.addf`) |
| which op shape carries it | `ComputeStep.surface` | `BARE` / `REDUCE` / `GENERIC` |
| what a generic must state | `ComputeStep.indexing` | iterator types + one projection per operand, `None` unless `GENERIC` |

`Family` is deleted — reducing is exactly `kind is COMBINER`, and the recipe-vs-spec agreement check becomes `(recipe.kind is COMBINER) != spec.is_reduction`. Dispatch is an explicit `match step.surface`, so the two wrapper methods that existed only to pick a shape are gone.

The surface is chosen in the plan, from the derived nest:

```
add            identity maps, all parallel                      -> BARE     linalg.add
sum  dim=0     identity in, identity-minus-reduced out          -> REDUCE   linalg.reduce
sum  dim=-1    (0,1,2) in, (1,3) out over a 4-dim nest          -> GENERIC  linalg.generic
```

The load-bearing part of that rule is that it tests the **input** map. `mlir::linalg::ReduceOp` derives its maps as `getMultiDimIdentityMap(rank).dropResults(dimensions)`, so it can only read its input with one loop per input dimension. On-stick reads three input dimensions from a four-dimension nest, so it falls through to `GENERIC`. Judged on the output alone it would look legal — `(1,3)` *is* an increasing subsequence of `(0,1,2,3)` — and the emitter would quietly build a rank-2 `linalg.reduce` that sums the wrong elements. There is a test pinned on exactly that.

#### Commits, each independently green

| commit | |
|---|---|
| `4ba4bff` | pins `ktir-mlir-frontend` and builds `memory_space` through `ktdp.MemorySpaceAttr` instead of parsing it from text |
| `acf2601` | splits the *shape* of an op out of the method that picks it (`_emit_bare` / `_emit_reduce`) |
| `d50e095` | dispatch on `Surface`; delete `Family` |
| `ce2c1df` | `reduction_indexing` replaces `reduced_axes`: derive the whole nest, not just which axes vanish |
| `eec4119` | `_emit_generic`, and the on-stick reduction as its first caller |
| `9bad1f7` | drops the symbolic-args prerequisite, which no longer holds |
| `860c6c0` | updates the two tests that pinned that prerequisite |

**The five pre-existing goldens are byte-identical across the first four commits.** That is the only claim those commits make — no emitted text moves. `eec4119` adds one new golden and changes none.

The last two commits are not about the surface. `9bad1f7` drops the `BUNDLE_SYMBOLIC_ARGS=0` prerequisite from `_check_ktir_device_prerequisites`: the backend takes a symbolic base address now, so demanding baked constants refused work that in fact compiles. `860c6c0` follows it into the tests that pinned that refusal — one inverted to assert the check stays quiet with symbolic args on, the other counting the two prerequisites that remain (`KTIR_DEVICE_MLIR`, and `dbo-opt` on `PATH`).

#### Testing

`tests/inductor/test_ktir_{emitter,validate,compile}.py` + `test_opspec_utils.py`: **126 passed, 34 subtests**.

The new golden was checked against the device-verified fixture `examples/sum/variants/1core_onstick_loopless/ktir/sum_onstick.ktir` in the test framework, item by item, rather than captured from the emitter and assumed correct: both indexing maps, the iterator types, the `outs` shape, the output view's sizes/strides, and that `outs` is a bare `tensor.empty` with no `linalg.fill`. All agree.

**On device**, against CPU eager (full output in the comment below):

| case | | |
|---|---|---|
| `sum-onstick` | `[256, 128]`, 1 core | **PASS**, max diff 0.6875 on values up to 71.6 |
| `sum-onstick-divided` | `[8192, 128]`, 32 cores | **PASS**, max diff 0.6875 on values up to 76.8 |

Work division does the right thing for this shape — the parallel axis is the rows and the reduced axis is left whole, so no cross-core combine is needed. Confirmed by calling the splitter directly rather than inferring it:

```
[256, 4096] dim=-1  ->  c0:32, c1:1     # rows split 32 ways, reduced axis undivided
[256, 4096] dim=0   ->  c1:32, c0:1     # for contrast, the nonstick case
```

#### Which issue(s) this PR is related to:

Related to #3380 (the OpSpec→KTIR series) and torch-spyre/dataflow-scheduler#37 (operations with stick reduction) and torch-spyre/dataflow-scheduler#36 (opaque operations) 

#### Does this PR introduce a user-facing change?

No. Everything is behind `TORCH_SPYRE_KTIR=1`; with it unset the SuperDSC path is selected and none of this is reached.

### CODE REVIEW

**Comment by fabianlim** [comment]
@dgrove-oss i decided to remove the spyreop commit first, and first try to merge the sum-onstick, since that is the one that has already been verified on device. the spyreop may take more time. Recording the previous commit 9778e6d5

**Review by dgrove-oss — DISMISSED** [review]
lgtm

**Comment by dgrove-oss** [comment]
The change to async_compile requires an execution code owner to approve.

File: pyproject.toml [hunk 1]
```diff
@@ -47,8 +47,12 @@ dynamic = ["version"]
 # out-of-band MLIR build (CMAKE_ARGS=-DMLIR_DIR=...), so it cannot be installed
 # unattended. The marker is required: it needs >=3.12, this project allows >=3.11,
 # and uv.lock must resolve for every version supported.
+#
+# The floor is ktir-mlir-frontend#61: `memory_view` builds the `memory_space`
+# attribute through `ktdp.MemorySpaceAttr`, and an earlier rev raises
+# AttributeError on it.
 ktir = [
-    "ktir-mlir-frontend @ git+https://github.com/torch-spyre/ktir-mlir-frontend.git@05e7a0ff814ebefb6e78450b1ec0e871c488c8bc ; python_version >= '3.12'",
+    "ktir-mlir-frontend @ git+https://github.com/torch-spyre/ktir-mlir-frontend.git@91455932139b219653f2396b0861b7aaf1961d6b ; python_version >= '3.12'",
```
Comment (dgrove-oss) [inline pyproject.toml]: Changes to pyproject.toml require a top-level code owner to approve.  If you think this will change frequently, there will be less friction in approving PRs if you figure out some other place to put this.

Comment (fabianlim) [reply pyproject.toml]: ok then maybe for now i will just pin this on main

Comment (fabianlim) [reply pyproject.toml]: sorry i cant do, i need a fixed pin. otherwise if I merge PRs on that ktir-mlir repo, then somebody else's PR will fail the lock check. 

The only other option is to remove it from uv.lock

Comment (fabianlim) [reply pyproject.toml]: @dgrove-oss how about this, i create a script in `tools/` to install the ktir-mlir dep, where I also pin the commit here. But in order to avoid the similar problem with pyproject.toml, I need to create a path in tooks and assign CODEOWNERS in there. 

**Edit**: put a commit [1b9a8c5](https://github.com/torch-spyre/torch-spyre/pull/4028/commits/1b9a8c54bb34bad43a3d721d0e22c1ea53785693) here

Alternatively i put the script in tools/ but i put the pin somewhere in the `torch-spyre/inductor/` path, but this is messy, if tools is considered to not require top-level approval then the first proposal is better

File: CODEOWNERS [hunk 2]
```diff
@@ -49,3 +49,9 @@
 /pytest.ini               @ashokponkumar @jabostian @anubhavjana @n-marion
 /pytest_models.ini        @ashokponkumar @jabostian @anubhavjana @n-marion
 /Makefile                 @ashokponkumar @jabostian @anubhavjana @n-marion
+
+# ktir-mlir-frontend pin and its installer.  Two files rather than /tools/, which
+# is otherwise CI's: which frontend revision the OpSpec->KTIR emitter is tested
+# against is a fact about /torch_spyre/_inductor/.
+/tools/ktir-mlir-frontend.pin  @dgrove-oss @avery-blanchard @tardieu @fabianlim @tnakaike
+/tools/install-ktir.sh         @dgrove-oss @avery-blanchard @tardieu @fabianlim @tnakaike
```
Comment (dgrove-oss) [inline CODEOWNERS]:
```suggestion
/tools/install-ktir.sh         @dgrove-oss @avery-blanchard @tardieu
```

File: CODEOWNERS [hunk 3]
```diff
@@ -49,3 +49,9 @@
 /pytest.ini               @ashokponkumar @jabostian @anubhavjana @n-marion
 /pytest_models.ini        @ashokponkumar @jabostian @anubhavjana @n-marion
 /Makefile                 @ashokponkumar @jabostian @anubhavjana @n-marion
+
+# ktir-mlir-frontend pin and its installer.  Two files rather than /tools/, which
+# is otherwise CI's: which frontend revision the OpSpec->KTIR emitter is tested
+# against is a fact about /torch_spyre/_inductor/.
+/tools/ktir-mlir-frontend.pin  @dgrove-oss @avery-blanchard @tardieu @fabianlim @tnakaike
```
Comment (dgrove-oss) [inline CODEOWNERS]:
```suggestion
/tools/ktir-mlir-frontend.pin  @dgrove-oss @avery-blanchard @tardieu
```

**Review by dgrove-oss — DISMISSED** [review]
Summary

Separates three previously-tangled facts in the KTIR emitter — Recipe.kind (what a binding is), ComputeStep.surface (which linalg shape carries it: BARE/REDUCE/GENERIC), and ComputeStep.indexing (explicit maps/iterators, only for GENERIC) — replacing the old Family enum/dispatch. This lets an on-stick (dim=-1) reduction be emitted correctly as linalg.generic instead of an incorrect linalg.reduce, since mlir::linalg::ReduceOp can only express an identity-mapped input with dropped dimensions, and the on-stick case reads 3 input dims through a 4-dim iteration nest. Also drops the now-unneeded BUNDLE_SYMBOLIC_ARGS=0 prerequisite, switches memory_space from a hand-parsed text attribute to a verifier-checked builder API, and moves the ktir-mlir-frontend pin out of pyproject.toml/uv.lock into a new tools/install-ktir.sh + tools/ktir-mlir-frontend.pin, narrowly CODEOWNERS-gated.

BLOCKERs

None found.

SUGGESTIONs

- _emit_generic's body(*args) drops the last arg when not reducing (args[:-1]), relying on linalg.generic's block-arg convention rather than asserting arity first — a future two-output GENERIC step would fail with a confusing MLIR-construction error instead of a clear assertion (ktir.py::_emit_generic). Minor, no such step exists yet.

QUESTIONs

- PAYLOAD binding kind is only exercised via a synthetic recipe in tests — tracking issue for the first real spyreop intrinsic?
- The on-stick golden explicitly says the emitted kernel isn't compilable yet (lane-reduction lowering doesn't exist). Is that lowering a planned follow-up, or is this intentionally emitter-only scaffolding?

Verdict

Needs a fix to CODEOWNERS (only people with write access to the repo can be code owners).   Otherwise looks good.

File: pyproject.toml (hunk 1 above)
Comment (dgrove-oss) [reply pyproject.toml]: The changes you made to put it in tools handle this reasonably.

**Comment by fabianlim** [comment]
@dgrove-oss updated CODEOWNERS

**Review by dgrove-oss — APPROVED** [review]
lgtm

**Review by mudhakar — APPROVED** [review]
lgtm

**Review by thoangtrvn — APPROVED** [review]
